File: apps/mobile/modules/audio-waveform/android/src/main/java/expo/modules/audiowaveform/AudioWaveformModule.kt
package expo.modules.audiowaveform

import expo.modules.kotlin.Promise
import expo.modules.kotlin.modules.Module
import expo.modules.kotlin.modules.ModuleDefinition
import java.util.concurrent.Executors

class AudioWaveformModule : Module() {
  // `AsyncFunction` runs on a dispatcher shared by every Expo module in the app,
  // so a multi-second decode there would stall unrelated native calls (the
  // library indexer's metadata reads, for one). Own the thread instead — which
  // also serializes decodes, keeping a single decoder instance alive at a time.
  private val executor = Executors.newSingleThreadExecutor { r ->
    Thread(r, "wavio-waveform").apply { isDaemon = true }
  }

  override fun definition() = ModuleDefinition {
    Name("AudioWaveform")

    AsyncFunction("extractPeaks") { uri: String, buckets: Int, maxDurationMs: Long, promise: Promise ->
      executor.execute {
        try {
          val result =
            PeakExtractor.extract(appContext.reactContext, uri, buckets, maxDurationMs)
          promise.resolve(
            mapOf(
              "peaks" to result.peaks.toList(),
              "durationMs" to result.durationMs,
            ),
          )
        } catch (e: CodecFailure) {
          promise.reject(e)
        } catch (e: Exception) {
          promise.reject(WaveformDecodeException("${e.javaClass.simpleName}: ${e.message}"))
        }
      }
    }

    OnDestroy {
      executor.shutdownNow()
    }
  }
}

File: apps/mobile/modules/audio-waveform/ios/AudioWaveformModule.swift
import ExpoModulesCore

public class AudioWaveformModule: Module {
  // `AsyncFunction` otherwise runs on a queue shared by every Expo module, so a
  // multi-second decode there would stall unrelated native calls. Owning the
  // queue also serializes decodes, keeping one reader alive at a time.
  private let queue = DispatchQueue(label: "app.wavio.waveform", qos: .utility)

  public func definition() -> ModuleDefinition {
    Name("AudioWaveform")

    AsyncFunction("extractPeaks") { (uri: String, buckets: Int, maxDurationMs: Int64) -> [String: Any] in
      let result = try PeakExtractor.extract(
        uri: uri, buckets: buckets, maxDurationMs: maxDurationMs)
      return ["peaks": result.peaks, "durationMs": result.durationMs]
    }
    .runOnQueue(queue)
  }
}
